mongoimport/common.go

```diff
@@ -380,13 +544,37 @@ func tokensToBSON(colSpecs []ColumnSpec, tokens []string, numProcessed uint64, i
 }
 
 // validateFields takes a slice of fields and returns an error if the fields
-// are invalid, returns nil otherwise
-func validateFields(fields []string) error {
-	fieldsCopy := make([]string, len(fields), len(fields))
-	copy(fieldsCopy, fields)
-	sort.Sort(sort.StringSlice(fieldsCopy))
+// are invalid, returns nil otherwise. Fields are invalid in the following cases:
+//
+//     (1). A field contains an invalid series of characters
+//     (2). Two fields are the same (e.g. a,a)
+//     (3). One field implies there is a value, another implies there is a document (e.g. a,a.b)
+//
+// In the case that --useArrayIndexFields is set, fields are also invalid in the following cases:
+//
+//     (4). One field implies there is a value, another implies there is an array (e.g. a,a.0). This
+//          check is covered by the check for number 2.
+//     (5). One field implies that there is a document, another implies there is an array.
+//          (e.g. a.b,a.0 or a.b.c,a.0.c)
+//     (6). The indexes for an array don't start from 0 (e.g. a.1,a.2)
+//     (7). Array indexes are out of order (e.g. a.1,a.0 or a.0.b,a.1,a.0.c)
+//     (8). An array is missing an index (e.g. a.0,a.2)
+func validateFields(inputFields []string, useArrayIndexFields bool) error {
+	fields := make([]FieldInfo, len(inputFields))
```

Comment:
As we discussed in the meeting, we're going to try an approach that builds a trie with embedded arrays (where necessary). It will be able to handle this linearly as long as we require that array indexes always be specified in order by the user.